### DISCUSSION TITLE
Questions about profile aware files

### DISCUSSION THREAD

**Discussion by TopSecretAtWork** [quarkusio/quarkus#25332] [Q&A] [answered]
Hello Everyone,

We have a question about the [profile aware property files](https://quarkus.io/guides/config-reference#profile-aware-files). We do have separate files for different environments (prod, staging, local, ...) and we activate the different profiles in their corresponding environments.

From our point of view, we have some strange behavior when running the application from the command line and we wanted to ask if someone could clarify that with us. We have created a small [repo](https://github.com/TopSecretAtWork/quarkus-profile-aware-files) to reproduce that behavior.

About the setup:

* openjdk 17.0.3
* mvn 3.8.1
* Quarkus 2.8.2.Final

application-common.properties:
```
##################################################################################
#                                     COMMON                                     #
##################################################################################


##### INTERVAL #####
send.interval.in.sec=30
```

application-staging.properties:
```
##################################################################################
#                                    STAGING                                     #
##################################################################################

##### PARENT #####
quarkus.config.profile.parent=common



##### SERVICES #####
services.test.SERVICE1=http://localhost:8080/health
```

application-prod.properties
```
##################################################################################
#                                      PROD                                      #
##################################################################################

##### PARENT #####
quarkus.config.profile.parent=common



##### SERVICES #####
services.test.SERVICE1=http://localhost:8080/health
services.test.SERVICE2=http://localhost:8081/health
services.test.SERVICE3=http://localhost:8082/health


value.in.prod.only=true
```

When we start the application with the `staging` profilein our IDE (IntelliJ) everything works as we would expect it. Only the `SERVICE1` value is loaded and with the `prod` profile, all three services are loaded.

When we package and start the application via the CLI in both profiles (`staging` and `prod`) all three service values are loaded. It seems that quarkus loads the `prod` properties, regardless of which profile was passed, and provides those values via a `io.smallrye.config.PropertiesConfigSource` named `PropertiesConfigSource[source=Specified default values]`.

Here is the log output when we run the sample project with `-Dquarkus.profile=staging` from the command line

```java
2022-05-03 10:50:12,537 INFO  [com.tes.con.con.use.Startup] (main) Config property names: '[services.test.SERVICE1, services.test.SERVICE2, services.test.SERVICE3]'
2022-05-03 10:50:12,537 INFO  [com.tes.con.con.use.Startup] (main) name: 'services.test.SERVICE1' -> value: 'http://localhost:8080/health'; (source: ConfigValue{name='services.test.SERVICE1', value='http://localhost:8080/health', rawValue='http://localhost:8080/health', profile='null', configSourceName='PropertiesConfigSource[source=jar:file:/tmp/quarkus-app/app/configtest-main-0.1.0-SNAPSHOT.jar!/application-staging.properties]', configSourceOrdinal=253, configSourcePosition=3, lineNumber=-1})
2022-05-03 10:50:12,538 INFO  [com.tes.con.con.use.Startup] (main) name: 'services.test.SERVICE2' -> value: 'http://localhost:8081/health'; (source: ConfigValue{name='services.test.SERVICE2', value='http://localhost:8081/health', rawValue='http://localhost:8081/health', profile='null', configSourceName='PropertiesConfigSource[source=Specified default values]', configSourceOrdinal=-2147483548, configSourcePosition=7, lineNumber=-1})
2022-05-03 10:50:12,538 INFO  [com.tes.con.con.use.Startup] (main) name: 'services.test.SERVICE3' -> value: 'http://localhost:8082/health'; (source: ConfigValue{name='services.test.SERVICE3', value='http://localhost:8082/health', rawValue='http://localhost:8082/health', profile='null', configSourceName='PropertiesConfigSource[source=Specified default values]', configSourceOrdinal=-2147483548, configSourcePosition=7, lineNumber=-1})
2022-05-03 10:50:12,538 INFO  [com.tes.con.con.use.Startup] (main) Config profiles: [staging, common, common]
2022-05-03 10:50:12,539 INFO  [com.tes.con.con.use.Startup] (main) Default config source (class: class io.smallrye.config.PropertiesConfigSource): {value.in.prod.only=true, send.interval.in.sec=30, quarkus.config.profile.parent=common, services.test.SERVICE1=http://localhost:8080/health, services.test.SERVICE2=http://localhost:8081/health, services.test.SERVICE3=http://localhost:8082/health}
```
We can see that the `SERVICE1` value is loaded from the `application-staging.properties` file and the other two values are loaded from the default `PropertiesConfigSource`. The last log entry shows the default source, which is coming from the `application-prod.properties` file.

And here is now the log output when we run the sample project with `-Dquarkus.profile=prod` from the command line

```java
2022-05-03 10:56:05,750 INFO  [com.tes.con.con.use.Startup] (main) Config property names: '[services.test.SERVICE1, services.test.SERVICE2, services.test.SERVICE3]'
2022-05-03 10:56:05,750 INFO  [com.tes.con.con.use.Startup] (main) name: 'services.test.SERVICE1' -> value: 'http://localhost:8080/health'; (source: ConfigValue{name='services.test.SERVICE1', value='http://localhost:8080/health', rawValue='http://localhost:8080/health', profile='null', configSourceName='PropertiesConfigSource[source=jar:file:/tmp/quarkus-app/app/configtest-main-0.1.0-SNAPSHOT.jar!/application-prod.properties]', configSourceOrdinal=253, configSourcePosition=3, lineNumber=-1})
2022-05-03 10:56:05,751 INFO  [com.tes.con.con.use.Startup] (main) name: 'services.test.SERVICE2' -> value: 'http://localhost:8081/health'; (source: ConfigValue{name='services.test.SERVICE2', value='http://localhost:8081/health', rawValue='http://localhost:8081/health', profile='null', configSourceName='PropertiesConfigSource[source=jar:file:/tmp/quarkus-app/app/configtest-main-0.1.0-SNAPSHOT.jar!/application-prod.properties]', configSourceOrdinal=253, configSourcePosition=3, lineNumber=-1})
2022-05-03 10:56:05,751 INFO  [com.tes.con.con.use.Startup] (main) name: 'services.test.SERVICE3' -> value: 'http://localhost:8082/health'; (source: ConfigValue{name='services.test.SERVICE3', value='http://localhost:8082/health', rawValue='http://localhost:8082/health', profile='null', configSourceName='PropertiesConfigSource[source=jar:file:/tmp/quarkus-app/app/configtest-main-0.1.0-SNAPSHOT.jar!/application-prod.properties]', configSourceOrdinal=253, configSourcePosition=3, lineNumber=-1})
2022-05-03 10:56:05,751 INFO  [com.tes.con.con.use.Startup] (main) Config profiles: [prod, common, common]
2022-05-03 10:56:05,752 INFO  [com.tes.con.con.use.Startup] (main) Default config source (class: class io.smallrye.config.PropertiesConfigSource): {value.in.prod.only=true, send.interval.in.sec=30, quarkus.config.profile.parent=common, services.test.SERVICE1=http://localhost:8080/health, services.test.SERVICE2=http://localhost:8081/health, services.test.SERVICE3=http://localhost:8082/health}
```
Now all values, as expected, are loaded from the `application-prod.properties` file.

We would expect that in the case of the staging profile we just load the `SERVICE1` value and not all of them. As mentioned before this does not happen when we start the application through the IDE, in this case, the default source is just empty as shown below

```java
2022-05-03 10:58:56,320 INFO  [com.tes.con.con.use.Startup] (Quarkus Main Thread) Config property names: '[services.test.SERVICE1]'
2022-05-03 10:58:56,321 INFO  [com.tes.con.con.use.Startup] (Quarkus Main Thread) name: 'services.test.SERVICE1' -> value: 'http://localhost:8080/health'; (source: ConfigValue{name='services.test.SERVICE1', value='http://localhost:8080/health', rawValue='http://localhost:8080/health', profile='null', configSourceName='PropertiesConfigSource[source=file:/home/pwanko/NTS/DEV/Code/shitfuck/configtest/configtest-main/target/classes/application-staging.properties]', configSourceOrdinal=253, configSourcePosition=4, lineNumber=-1})
2022-05-03 10:58:56,321 INFO  [com.tes.con.con.use.Startup] (Quarkus Main Thread) Config profiles: [staging, common, common]
2022-05-03 10:58:56,322 INFO  [com.tes.con.con.use.Startup] (Quarkus Main Thread) Default config source (class: class io.smallrye.config.PropertiesConfigSource): {}
```

We tried to debug this but all we could find so far was that the default `PropertiesConfigSource` with the default values is passed to the  method `public SmallRyeConfigBuilder withSources(ConfigSource... configSources)` in the class `io.smallrye.config.SmallRyeConfigBuilder`. The calls before seem to be generated and we weren't able to debug them, so we don't know exactly how and when the default properties are loaded. Maybe someone can point us in the right direction, please?

The current "workaround" is to keep the `application.properties` file empty and to rename the `application-prod.properties` file (e.g.: to `application-live.properties`). But we do not know for sure if our setup here is completely wrong or if this is a bug.

So we wanted to ask the community if someone could tell us please if this is working as designed and our setup is wrong or if this is a bug?

Kind Regards
Philipp

**Comment by radcortez** [comment, marked as answer]
Hi Philipp,

Yes, this is expected and working as intended. I know it can be a little confusing. Let me explain.

The Quarkus application is built with the `prod` profile, and we record the configuration values in a minimal ordinal `ConfigSource`. This action intends to prevent the application startup from failing because the configuration required during the build may be absent from runtime. Another reason is to remove the file reading from the native image binary.

So, in this case, because you have more configuration set up for the `prod` profile (service 2 and service 3), when you activate the `staging`, which only contains service 1, you will still see service 2 and 3. You also noticed that the values are coming from `'PropertiesConfigSource[source=Specified default values]', configSourceOrdinal=-2147483548`, where we record the values during the build.

My recommendation is to just empty the values in the `staging` profile like this:

```properties
##### SERVICES #####
services.test.SERVICE1=http://localhost:8080/health
services.test.SERVICE2=
services.test.SERVICE3=
```

This will override the recorded `prod` values and remove the configuration.

Another option is to use `quarkus.config.locations` as a system property or environment variable and point to an additional config file or folder that is only available during runtime (so the values are not recorded).

I hope this makes sense.

**Reply by harel-biton** [reply]
Hi @radcortez,

Sorry for reviving this old thread, but I recently ran into this behavior myself while working with Quarkus. I have to admit I found it quite unintuitive, and in some cases potentially dangerous.

From what I can see, even some runtime (not only build-time) configuration values are effectively hardcoded during the build and may later be used as fallbacks at runtime. If I’m not mistaken, at least some of these values end up in `io.quarkus.runtime.generated.RunTimeConfigCustomizer#runtimeValues`.

In your earlier comment, you mentioned two motivations for this behavior:
1. Preventing application startup failures when configuration required during the build is missing at runtime.
2. Avoiding file reads from within the native image binary.

Regarding (1):
I’m wondering why preventing startup failure is necessarily the desired outcome in this case. If configuration is part of the application contract and it’s missing at runtime, failing fast seems like a legitimate, and sometimes even preferable, outcome. Silent fallback to build-time values can mask misconfiguration and potentially lead to unexpected behavior in production. Even if avoiding startup failures is valuable in some scenarios, I’m not sure it outweighs the risks introduced by implicitly carrying over configuration from a different profile.

Regarding (2):
If the goal is to avoid file reads in native images, would it make sense to hardcode the values together with their originating profile and only apply them at runtime if they match the currently active profile? That would still avoid file I/O while preventing cross-profile leakage of configuration.
I’d be very interested to understand whether this behavior is considered fundamental to Quarkus’ design, or if there might be room to make it more explicit or profile-aware in the future.

Thanks again for your previous explanations - they were very helpful in understanding what’s going on under the hood (I ended up spending a few hours investigating this).

**Reply by radcortez** [reply]
> I’m wondering why preventing startup failure is necessarily the desired outcome in this case. If configuration is part of the application contract and it’s missing at runtime, failing fast seems like a legitimate, and sometimes even preferable, outcome. Silent fallback to build-time values can mask misconfiguration and potentially lead to unexpected behavior in production. Even if avoiding startup failures is valuable in some scenarios, I’m not sure it outweighs the risks introduced by implicitly carrying over configuration from a different profile.

Well, we believe that Quarkus should always be able to start. So, we record static configuration (as in properties files, yaml files, etc), and we avoid build-time environment configuration like system properties or env vars. The configuration is already available in the application, unless someone did something destructive and started to remove files from the built app :)

Now, we have multiple configuration phases. When running a Quarkus App, most notably, a build and runtime fixed config and a runtime config. The build and runtime fixed config means that the config is read and used at build time, and it is available for reading at runtime. Obviously, we cannot allow this configuration phase to be changed. The second phase, runtime config is the one that most users use. Yes, in this case, we record the configuration we can find at build time as a fallback default.

> If the goal is to avoid file reads in native images, would it make sense to hardcode the values together with their originating profile and only apply them at runtime if they match the currently active profile? That would still avoid file I/O while preventing cross-profile leakage of configuration.
> I’d be very interested to understand whether this behavior is considered fundamental to Quarkus’ design, or if there might be room to make it more explicit or profile-aware in the future.

We can consider exploring recording these fallback defaults as explicit `prod` config, but unfortunately, it would always be a big / bad breaking change for users running in non-prod profiles.

Also, can you share with me your specific use cases? That could help figure out something better. Thanks!

**Reply by harel-biton** [reply]
Here’s the specific issue I ran into, which might help illustrate the concern.

I’m using the quarkus-opentelemetry extension, with the following configuration:

**application.yaml**
(no OpenTelemetry config)

**application-prod.yaml**
```
otel:
  logs:
    enabled: true
    level: INFO
  exporter:
    otlp:
      logs:
        endpoint: https://some-listener-endpoint.com
        headers: "Authorization=Bearer ${api-key}"
```
**application-stg.yaml**
```
otel:
  enabled: false
```
(I also have a custom config source that resolves the api-key at startup)

During the build, the property `otel.exporter.otlp.logs.headers` gets recorded into `io.quarkus.runtime.generated.RunTimeConfigCustomizer#runtimeValues`.

When I run the app with the `stg` profile, the application still tries to resolve `${api-key}` and fails (the value is not available in that environment) - even though OpenTelemetry is disabled in that profile.

I’m not entirely sure whether this behavior originates from the extension (e.g., eager resolution) or from Quarkus itself resolving recorded runtime values regardless of actual usage.

The workaround I used was to override the property in application.yaml:
```
otel:
  exporter:
    otlp:
      logs:
        headers: "empty"
```
This prevents the resolution failure because it overrides the recorded prod value.

My concerns with the current "prod as runtime fallback" behavior:
1. Workaround is not very elegant - the base application.yaml can become cluttered with override values that are only there to neutralize unrelated profile configs.
2. Risk of silent misconfiguration - in my case, the app failed to start, so the issue was visible. But in other cases, the app might start successfully while using unintended configuration from another profile - which could be much harder to detect.
3. Non-intuitive behavior - even understanding Quarkus reasonably well, this behavior was surprising. It’s easy to forget that the prod profile can still influence runtime via recorded defaults even when it is not the active profile.

I understand the backward compatibility concerns, but perhaps this could be addressed with an opt-in build flag that records fallbacks as profile-specific (e.g., only applied when prod is active). This would allow teams to opt into stricter and more intuitive behavior without breaking existing applications.
This would also reduce the cognitive overhead - instead of developers needing to remember this behavior for every new configuration, it becomes a one-time, explicit project-level decision.

Would love to hear your thoughts.

**Reply by radcortez** [reply]
Thanks, let me experiment with this a bit and I'll get back to you.

**Reply by harel-biton** [reply]
@radcortez did you have a chance to check this?

**Reply by radcortez** [reply]
Not yet. Sorry. I'll see if I can look into it in the next couple of weeks.

**Reply by radcortez** [reply]
I had a look, and an alternative is to force the profile in the configuration section, such as:

```yaml
"%prod":
  quarkus:
    otel:
      exporter:
        otlp:
          logs:
            headers: "Authorization=Bearer ${api-key}"
```

This will force the recording into `prod`, instead of unprofiled.

**Reply by harel-biton** [reply]
Thanks for looking into this and for the suggestion @radcortez - adding `"%prod":` at the top of my `application-prod.yaml` solved the issue.

I do have a question though, purely out of curiosity and to make sure I'm not misunderstanding something fundamental.

My intuition was that properties declared inside `application-prod.yaml` would naturally be recorded as prod-scoped during the build - since the file itself is already associated with the `prod` profile by its name. Having to additionally write `"%prod":` inside that file felt somewhat redundant to me.

Am I missing something about how Quarkus approaches configuration scoping that makes this behavior intentional and expected? Or is this more of an edge case / design quirk that exists today partly for historical reasons, where changing it now would be a breaking change even if the current behavior isn't ideal?

If it turns out the current behavior is indeed surprising by design, it might be worth calling out explicitly in the documentation, as this is easy to miss and could quietly cause issues for other users too.

Either way, I really appreciate your time and patience on this, thank you.

**Reply by radcortez** [reply]
> My intuition was that properties declared inside `application-prod.yaml` would naturally be recorded as prod-scoped during the build - since the file itself is already associated with the `prod` profile by its name. Having to additionally write `"%prod":` inside that file felt somewhat redundant to me.

No, they are recorded as unprofiled, since the recording profile is `prod`. Adding the `%prod` is a trick there to prefix all properties with `%prod.`, so they still get applied in the `prod` profile, but are recorded preserving the `prod` prefix.

I'll see what I can do to improve this.

**Comment by TopSecretAtWork** [comment]
Hi Roberto,

thank you very much for the answer and the explanation.
I understand now, how this works.

For now we had a "workaround" based on your second suggestion. We moved the services in a separate profile aware file and removed them from the prod profile and just loaded the other profile with `-Dquarkus.profile`

We will change our current solution to your second suggestion and will provide those values with `quarkus.config.locations`.
Thank you again for your help.

Kind Regards
Philipp
